lib/compiler-singlepass/src/compiler.rs

```diff
@@ -85,82 +86,109 @@ impl Compiler for SinglepassCompiler {
+                        wasmer_compiler::FunctionReader::new(input.module_offset, input.data);
+                    let mut generator = FuncGen::new(
+                        module,
+                        module_translation,
+                        &self.config,
+                        &vmoffsets,
+                        &table_styles,
+                        i,
+                        calling_convention,
+                    )
+                    .map_err(to_compile_error)?;
 
-                let mut local_reader = reader.get_locals_reader()?;
-                for _ in 0..local_reader.get_count() {
-                    let (count, ty) = local_reader.read()?;
-                    // Overflows feeding a local here have most likely already been caught by the
-                    // validator, but it is possible that the validator hasn't been run at all, or
-                    // that the validator does not impose any limits on the number of locals.
-                    generator.feed_local(count, ty);
-                }
+                    let mut local_reader = reader.get_locals_reader()?;
+                    for _ in 0..local_reader.get_count() {
+                        let (count, ty) = local_reader.read()?;
+                        // Overflows feeding a local here have most likely already been caught by the
+                        // validator, but it is possible that the validator hasn't been run at all, or
+                        // that the validator does not impose any limits on the number of locals.
+                        generator.feed_local(count, ty);
+                    }
 
-                generator.emit_head().map_err(to_compile_error)?;
+                    generator.emit_head().map_err(to_compile_error)?;
 
-                let mut operator_reader = reader.get_operators_reader()?.into_iter_with_offsets();
-                while generator.has_control_frames() {
-                    let (op, pos) = operator_reader.next().unwrap()?;
-                    generator.set_srcloc(pos as u32);
-                    generator.feed_operator(op).map_err(to_compile_error)?;
-                }
+                    let mut operator_reader =
+                        reader.get_operators_reader()?.into_iter_with_offsets();
+                    while generator.has_control_frames() {
+                        let (op, pos) = tracing::info_span!("parsing-next-operator")
+                            .in_scope(|| operator_reader.next().unwrap())?;
+                        generator.set_srcloc(pos as u32);
+                        generator.feed_operator(op).map_err(to_compile_error)?;
+                    }
 
-                Ok(generator.finalize(&input))
+                    Ok(generator.finalize(&input))
+                })
             })
             .collect::<Result<Vec<CompiledFunction>, CompileError>>()?
-            .into_iter()
+            .into_iter() // TODO: why not just collect to PrimaryMap directly?
```

Comment:
IIRC there are some special requirements for `ParallelIterator`. I tried fixing this mess once, and gave up quickly ^^

IMO we should just make rayon an unconditional dependency soon.